File: backend/app/api/catalog.py
import json
import hashlib
from typing import Optional
from fastapi import APIRouter, Depends, Query, Response, HTTPException, status
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.database import get_db
from app.storage import get_storage
from app.schemas.catalog import SearchResponse
from app.services.search_service import SearchService

router = APIRouter(tags=["Public Catalog"])

@router.get("/catalog")
async def get_published_catalog(response: Response):
    """
    Serves the pre-published catalogue JSON file directly from storage with ETag and Cache-Control headers.
    Ultra-low latency, zero database queries per request.
    """
    storage = get_storage()
    try:
        data_bytes = await storage.get(settings.CATALOG_FILE_PATH)
        data_json = json.loads(data_bytes.decode("utf-8"))
        
        # Calculate ETag
        etag = f'"{hashlib.md5(data_bytes).hexdigest()}"'
        response.headers["ETag"] = etag
        response.headers["Cache-Control"] = "public, max-age=60, stale-while-revalidate=300"
        return data_json
    except FileNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="The catalogue has not been published yet. Please publish the catalogue from the Admin CMS."
        )

@router.get("/catalog/search", response_model=SearchResponse)
async def search_catalog(
    q: Optional[str] = Query(None, description="Search query matching show titles, episode titles, or categories"),
    category: Optional[str] = Query(None, description="Filter by genre/category"),
    language: Optional[str] = Query(None, description="Filter by audio/subtitle language (e.g. en, es, hi, ja)"),
    section: Optional[str] = Query(None, description="Filter by section (e.g. Trending Now, Sci-Fi Hits)"),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
):
    """
    Dynamic composite catalogue search matching show title, episode title, and category.
    All filters compose at the SQL database layer.
    """
    return await SearchService.search_catalog(
        db=db,
        q=q,
        category=category,
        language=language,
        section=section,
        limit=limit,
        offset=offset,
    )

File: viewer/src/api/catalog.ts
import { CatalogPayload, SearchResponse } from '../types';

const BASE_URL = ((import.meta as any).env?.VITE_API_BASE_URL || '').replace(/\/$/, '');

export const getFullMediaUrl = (url?: string | null): string => {
  if (!url) return '';
  if (url.startsWith('http://') || url.startsWith('https://')) return url;
  return `${BASE_URL}${url.startsWith('/') ? '' : '/'}${url}`;
};

export const catalogApi = {
  /**
   * Fetches the pre-published atomic catalogue JSON.
   * Ultra-fast, cached with ETag, zero DB hits.
   */
  async getCatalog(): Promise<CatalogPayload> {
    const res = await fetch(`${BASE_URL}/catalog`);
    if (!res.ok) {
      if (res.status === 404) {
        throw new Error('Catalogue has not been published yet. Please publish from the CMS Admin portal.');
      }
      throw new Error(`Failed to load catalogue (${res.status})`);
    }
    return res.json();
  },

  /**
   * Composite SQL-level search endpoint.
   * Matches show title, episode title, category, language, and section.
   */
  async searchCatalog(params: {
    q?: string;
    category?: string;
    language?: string;
    section?: string;
  }): Promise<SearchResponse> {
    const searchParams = new URLSearchParams();
    if (params.q) searchParams.set('q', params.q);
    if (params.category) searchParams.set('category', params.category);
    if (params.language) searchParams.set('language', params.language);
    if (params.section) searchParams.set('section', params.section);

    const res = await fetch(`${BASE_URL}/catalog/search?${searchParams.toString()}`);
    if (!res.ok) {
      throw new Error(`Search failed (${res.status})`);
    }
    return res.json();
  },
};

